File: LLM_Translation/Rust_programs_compiled/rotate.rs
use std::cmp;
use std::ptr;

struct Node {
    key: i32,
    left: *mut Node,
    right: *mut Node,
    height: i32,
}

impl Node {
    fn new(key: i32) -> *mut Node {
        Box::into_raw(Box::new(Node {
            key,
            left: ptr::null_mut(),
            right: ptr::null_mut(),
            height: 1,
        }))
    }
}

fn max(a: i32, b: i32) -> i32 {
    cmp::max(a, b)
}

fn height(node: *mut Node) -> i32 {
    if node.is_null() {
        return 0;
    }
    unsafe { (*node).height }
}

fn right_rotate(y: *mut Node) -> *mut Node {
    unsafe {
        let x = (*y).left;
        let t2 = (*x).right;

        // Perform rotation and update heights
        (*y).left = t2;
        (*y).height = max(height((*y).left), height((*y).right)) + 1;
        (*x).right = y;
        (*x).height = max(height((*x).left), height((*x).right)) + 1;

        // Return new root
        x
    }
}

fn get_balance(node: *mut Node) -> i32 {
    if node.is_null() {
        return 0;
    }
    unsafe { height((*node).left) - height((*node).right) }
}

fn left_rotate(x: *mut Node) -> *mut Node {
    unsafe {
        let y = (*x).right;
        let t2 = (*y).left;

        // Perform rotation and update heights
        (*x).right = t2;
        (*x).height = max(height((*x).left), height((*x).right)) + 1;
        (*y).left = x;
        (*y).height = max(height((*y).left), height((*y).right)) + 1;

        // Return new root
        y
    }
}

fn insert(node: *mut Node, key: i32) -> *mut Node {
    if node.is_null() {
        return Node::new(key);
    }

    unsafe {
        if key < (*node).key {
            (*node).left = insert((*node).left, key);
        } else if key > (*node).key {
            (*node).right = insert((*node).right, key);
        } else {
            return node;
        }

        (*node).height = 1 + max(height((*node).left), height((*node).right));

        let balance = get_balance(node);

        // Left Left Case
        if balance > 1 && key < (*(*node).left).key {
            return right_rotate(node);
        }

        // Right Right Case
        if balance < -1 && key > (*(*node).right).key {
            return left_rotate(node);
        }

        // Left Right Case
        if balance > 1 && key > (*(*node).left).key {
            (*node).left = left_rotate((*node).left);
            return right_rotate(node);
        }

        // Right Left Case
        if balance < -1 && key < (*(*node).right).key {
            (*node).right = right_rotate((*node).right);
            return left_rotate(node);
        }
    }

    node
}

fn main() {
    let mut root: *mut Node = ptr::null_mut();
    root = insert(root, 7);
}

File: verification/rotate.c
#include<stdio.h>
#include<stdlib.h>


// An AVL tree node
struct Node
{
    int key;
    struct Node *left;
    struct Node *right;
    int height;
};




// A utility function to get maximum of two integers
int max(int a, int b) {
    return (a > b)? a : b;
}
 
// A utility function to get height of the tree
int height(struct Node *N)
{
    if (N == NULL)
        return 0;
    return N->height;
}

struct Node *rightRotate(struct Node *y)
{
    struct Node *x = y->left;
    struct Node *T2 = x->right;
 
    // Perform rotation and Update Heights
    y->left = T2;
    y->height = max(height(y->left), height(y->right))+1;
    x->right = y;
    x->height = max(height(x->left), height(x->right))+1;
 
    // Return new root
    return x;
}
 
// Get Balance factor of node N
int getBalance(struct Node *N)
{
    if (N == NULL)
        return 0;
    return height(N->left) - height(N->right);
}

struct Node *leftRotate(struct Node *x)
{
    struct Node *y = x->right;
    struct Node *T2 = y->left;
 
    // Perform rotation and Update Heights
    x->right = T2;
    x->height = max(height(x->left), height(x->right))+1;
    y->left = x;
    y->height = max(height(y->left), height(y->right))+1;
 
    // Return new root
    return y;
}

struct Node* insert(struct Node* node, int key)
{
    /* 1.  Perform the normal BST rotation */
    if (node == NULL)
        return(newNode(key));
 
    if (key < node->key)
        node->left  = insert(node->left, key);
    else if (key > node->key)
        node->right = insert(node->right, key);
    else // Equal keys not allowed
        return node;
 
    /* 2. Update height of this ancestor node */
    node->height = 1 + max(height(node->left),
                           height(node->right));
 
    /* 3. Get the balance factor of this ancestor
          node to check whether this node became
          unbalanced */
    int balance = getBalance(node);
 
    // If this node becomes unbalanced, then there are 4 cases
 
    // Left Left Case
    if (balance > 1 && key < node->left->key)
        return rightRotate(node);
 
    // Right Right Case
    if (balance < -1 && key > node->right->key)
        return leftRotate(node);
 
    // Left Right Case
    if (balance > 1 && key > node->left->key)
    {
        node->left =  leftRotate(node->left);
        return rightRotate(node);
    }
 
    // Right Left Case
    if (balance < -1 && key < node->right->key)
    {
        node->right = rightRotate(node->right);
        return leftRotate(node);
    }
 
    /* return the (unchanged) node pointer */
    return node;
}

int main()
{
  struct Node *root = NULL;
  root = insert(root, 7);

}